### ISSUE TITLE
HoareOptimizer pass should not keep internal state

### ISSUE THREAD

**Issue by 1ucian0** [Qiskit/qiskit-terra#4999]
The `HoareOptimizer` pass is storing internal state and that creates invalid results when run is called more than once.

```python
from qiskit import *
from qiskit.transpiler.passes import HoareOptimizer
from qiskit.converters import circuit_to_dag, dag_to_circuit

circuit1 = QuantumCircuit(2)
circuit1.z(0)
circuit1.h(1)
circuit1.z(1)

circuit2 = QuantumCircuit(2)
circuit2.z(1)
circuit2.h(0)
circuit2.z(0)

pass_ = HoareOptimizer()

result1 = pass_.run(circuit_to_dag(circuit1))
print(dag_to_circuit(result1))

result2 = pass_.run(circuit_to_dag(circuit2))
print(dag_to_circuit(result2))
```
```
q_0: ──────────
     ┌───┐┌───┐
q_1: ┤ H ├┤ Z ├
     └───┘└───┘
     ┌───┐
q_0: ┤ H ├
     └───┘
q_1: ─────
```

Notice that the result for `circuit2` is incorrect. The expected result can be obtain with a fresh pass instance:
```python
pass_ = HoareOptimizer()

result2 = pass_.run(circuit_to_dag(circuit2))
print(dag_to_circuit(result2))
```
```
     ┌───┐┌───┐
q_0: ┤ H ├┤ Z ├
     └───┘└───┘
q_1: ──────────
```

**Comment by pierrepocreau** [comment]
Hey, I'd like to try and solve this issue.

**Comment by Arijit02** [comment]
Hey! Can I check out this issue?

**Comment by 1ucian0** [comment]
Closed via #5138
